src/ol/source/Vector.js

```diff
@@ -495,32 +495,45 @@ class VectorSource extends Source {
         }
       }
     );
-    collection.addEventListener(
-      CollectionEventType.ADD,
-      /**
-       * @param {import("../Collection.js").CollectionEvent<import("../Feature.js").default<Geometry>>} evt The collection event
-       */
-      (evt) => {
-        if (!modifyingCollection) {
-          modifyingCollection = true;
-          this.addFeature(evt.element);
-          modifyingCollection = false;
-        }
+    /**
+     * @param {import("../Collection.js").CollectionEvent<import("../Feature.js").default<Geometry>>} evt The collection event
+     */
+    const handleCollectionAdd = (evt) => {
+      if (!modifyingCollection) {
+        modifyingCollection = true;
+        this.addFeature(evt.element);
+        modifyingCollection = false;
       }
-    );
+    };
+
+    /**
+     * @param {import("../Collection.js").CollectionEvent<import("../Feature.js").default<Geometry>>} evt The collection event
+     */
+    const handleCollectionRemove = (evt) => {
+      if (!modifyingCollection) {
+        modifyingCollection = true;
+        this.removeFeature(evt.element);
+        modifyingCollection = false;
+      }
+    };
+
+    collection.addEventListener(CollectionEventType.ADD, handleCollectionAdd);
     collection.addEventListener(
       CollectionEventType.REMOVE,
-      /**
-       * @param {import("../Collection.js").CollectionEvent<import("../Feature.js").default<Geometry>>} evt The collection event
-       */
-      (evt) => {
-        if (!modifyingCollection) {
-          modifyingCollection = true;
-          this.removeFeature(evt.element);
-          modifyingCollection = false;
-        }
-      }
+      handleCollectionRemove
     );
+
+    this.disposeTasks_.push(() => {
+      collection.removeEventListener(
+        CollectionEventType.ADD,
+        handleCollectionAdd
+      );
+      collection.removeEventListener(
+        CollectionEventType.REMOVE,
+        handleCollectionRemove
+      );
+    });
```

Comment:
This can be solved in a more elegant way with listener keys:
```js
this.listenerKeys_ = [
  listen(collection, CollectionEventType.ADD, ...);
  listen(collection, CollectionEventType.REMOVE, ...);
];
```
to create the listener keys, and then in `dispose()`:
```js
this.listenerKeys_.forEach(key => unlistenByKey(key));
```